_overviews/overview/a-taste-of-scala.md

```diff
@@ -0,0 +1,1329 @@
+
+```scala
+// define a case class
+case class Person(
+  name: String, 
+  vocation: String
+)
+
+// create an instance of the case class
+val p = Person("Reginald Kenneth Dwight", "Singer")
+
+// a good default toString method
+p                // Person = Person(Reginald Kenneth Dwight,Singer)
+
+// can access its fields, but they are immutable
+p.name           // "Reginald Kenneth Dwight"
+p.name = "Joe"   // error: can’t reassign a val field
+
+// when you need to make a change, use the `copy` method
+// to “update as you copy”
+val p2 = p.copy(name = "Elton John")
+p2               // Person = Person(Elton John,Singer)
+```
+
+See the Data Modeling sections of this Overview and the Reference documentation for many more details on case classes.
+
+
+
+## Scala methods
+
+<!--
+TODO: review Jonathan’s notes/categories about methods
+-->
+
+Scala classes, case classes, traits, enums, and objects can all contain methods. The general method syntax looks like this:
+
+```scala
+def methodName(param1: Type1, param2: Type2): ReturnType = 
+  // the method body
+  // goes here
+```
+
+Here are a few examples of that syntax:
+
+```scala
+def sum(a: Int, b: Int): Int = a + b
+def concatenate(s1: String, s2: String): String = s1 + s2
+```
+
+You don’t have to declare a method’s return type, so you can write those methods like this, if you prefer:
+
+```scala
+def sum(a: Int, b: Int) = a + b
+def concatenate(s1: String, s2: String) = s1 + s2
+```
+
+This is how you call those methods:
+
+```scala
+val x = sum(1, 2)
+val y = concatenate("foo", "bar")
+```
+
+Here’s an example of a multiline method:
+
+```scala
+def getStackTraceAsString(t: Throwable): String =
+  val sw = new StringWriter
+  t.printStackTrace(new PrintWriter(sw))
+  sw.toString
+```
+
+There are more things you can do with methods, such as providing default values for method parameters and using named parameters when calling methods. Those featurs are shown in the Data Modeling section of this Overview and in the Reference documentation.
+
+
+
+## First-class functions
+
+Scala has first-class support for functional programming, and included in that is first-class support for everything related to functions:
+
+- Top-level functions
+- Functions in objects
+- Lambdas
+- Higher-order functions (HOFs)
+
+<!-- TODO: when to say “function” vs when to say “method” -->
+<!-- TODO: show the `val` function syntax? -->
+
+
+###  Top-level functions
+
+Functions don’t have to be in the body of a class, enum, or object. They can also exist on their own as a *top-level function*. For instance, when this code is saved in a file, the functions `add` and `double` would be called top-level functions because they exist outside the scope of a class, enum, or object:
+
+```scala
+package foo
+
+def add(a: Int, b: Int): Int = a + b
+def double(a: Int): Int = 2 * a
+
+@main def topLevelFunctions() =
+  val x = add(1, 2)
+  val y = double(x)
+  println(s"y = $y")
+```
+
+
+### Functions in objects
```

Comment:
This example illustrates methods, we could move it up to the “methods” section?

Reply:
Yes, I’ll move it up there.